accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+##### What goes on the public Rekor ledger?
+Only public npm packages will be signed and published to the public [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview) ledger by default.
+
+Privately scoped packages, or packages from private repositories, will not be signed or published to the public ledger. This will be determined by interrogating the npm registry at publish time. Users will be able to override this similar to the `cosign --force` command by passing an argument to `npm publish`, e.g. `--force-build-provenance`.
+
+The entry will be stored and retrieved using the shasum of the package tarball. This will make sure we don't find any signatures for a  deleted package that has been resurrected with the same name and version.
+
+There will be no way of deleting an entry if this was published unintentionally to the ledger as it's immutable by design to prevent tampering. The practical consequence of publishing to the ledger is that information about your package (name, version), repository (commit, branch, owner) and workflow (name, id) are uploaded to the public ledger. A possible workaround would be to rename the leaked repo, make it private and delete the old one.
+
+##### Detailed steps to publish
+Interaction between the npm CLI tool running in CI, the npm registry, and the Sigstore components Fulcio and Rekor:
+
+```mermaid
+  sequenceDiagram
+    participant npm_cli
+    participant fulcio
+    participant rekor
+    participant npm_registry
+    npm_cli->>fulcio: Request signing certificate
+    fulcio->>npm_cli: Signing certificate
+    npm_cli->>rekor: Upload signed build provenance attestation
+    rekor->>npm_cli: Rekor entry
+    npm_cli->>npm_registry: Publish package and upload Rekor entry
+    npm_registry->>fulcio: Request signing certificate
+    fulcio->>npm_registry: Signing certificate
+    npm_registry->>rekor: Upload signed release attestations
+    rekor->>npm_registry: Rekor entry
+    npm_registry->>npm_registry: Attach rekor entry to published package
+    npm_registry->>npm_cli: Success
+```
+
+1. `npm publish --build-provenance` detects execution within a supported CI/CD environment and initializes token exchange to get a JWT id token from the CI/CD system's OIDC identity provider.
+2. npm CLI creates a new ephemeral key pair, requests a certificate from Fulcio (CA) and authenticates via the JWT id token.
+3. A SLSA build provenance attestation is created based on information available within the JWT and CI system and signs it with the ephemeral key.
+4. npm CLI uploads the completed attestation record on Rekor for public packages.
+5. npm registry creates a release attestation, stating that the version has been accepted and published, and places the record on Rekor for public packages.
```

Comment:
(Maybe I missed this) Can you motivate why we want the registry to upload a release attestation in addition to the artifact attestation? What does this second attestation add?

Reply:
Yes the way I see it is that the artifact or build attestation links the package an authenticated build identity. The release attestation will be created by the implementing registry, stating that the user who is publishing this package is in fact authorized to do so and the package has been published.

This links the build identity to an authorized publish. In future if we supported staged releases for example, the release attestation would only be created when the staged release was approved. It also creates an audit trail for public npm publishes, potentially allowing prudent organisations to monitor the log for any publishes to packages or org scopes they own.

Reply:
Would you consider publishing all events that occur for a package to the log? For example, in addition to publishing the release, what about yanking? What about updating critical metadata (such as who the maintainer is)?

Reply:
When we looked at this question for RubyGems (before we started working with other ecosystems through the OpenSSF's Securing Software Repos WG), we identified these operations as candidates for recording to the transparency log:

* Registration
* Sign during build
* Push
* Publish
* Yank
* Owner added
* Owner removed
* Owner changed email address
* Owner enabled MFA
* Owner disabled MFA
* Endorsement by non-owner
* Ownership call
* Ownership request
* Ownership transfer
* Deletion
* Resurrection of name

Reply:
More generally I see publishing a rich set of events to the log as creating a safe independent history of sensitive events. This provides the possibility of detection when the repo DB or blob bucket are compromised, because it would be out of sync with the transparency log.

Reply:
> Would you consider publishing all events that occur for a package to the log?

Yes would like to explore this more. The obvious one that jumps out is publishing an entry for deletions/yanks and malware takedowns.

It's probably a good idea to go slow adding these to make sure we have a good story for knowing what to do with it once we have it, e.g. during verification, and making sure we don't start spamming Rekor with too many events.

Reply:
@feelepxyz My ask is that we coordinate a bit on the schema for such events through the Securing Software Repos WG. One of my north stars is for my colleagues in AppSec to only need to install a single log monitor for all the ecosystems we rely on.

Reply:
> @feelepxyz My ask is that we coordinate a bit on the schema for such events through the Securing Software Repos WG.

Yes sounds good to me 👍

Reply:
@jchestershopify I added a section at the bottom about publishing additional events and aligning with the securing repos wg 👍